### ISSUE TITLE
Inference problem?

### ISSUE THREAD

**Issue by terrychenism** [HyeonwooNoh/DecoupledNet#1]
I think there exit some problem in inference process. You forward once to get a cls score and bp to pool5 and then forward to whole deconv process. You have calculate softmax when you get the result. But I think the softmax score doesn't work for segmentation. e.g. 21 channel, 1st channel is background and 7th channel contain true object. The probability of true object seems smaller than that of background channel when calculate by softmax, so that the result map is all 0(max of all channels). How do you solve this problem?

Thanks.

**Comment by HyeonwooNoh** [comment]
Segmentation network of DecoupledNet produce binary segmentation score, so we need to forward propagate several times to construct segmentation for whole class. The meaning of forward and background score of segmentation network is based on which classification score did you backpropagate. If you backpropagate every detected forward ground scores and then use it for segmentation, generated background score of binary segmentation is background score of overall semantic segmentation. To obtain score for foreground class, you have to backpropagate the score of one specific class (set other scores to zero). Forwarground score of obtained binary segmentation using this feature is segmentation score of the specific class. After obtaining overall segmentation through this way, we simply normalize score in the channel direction.

**Comment by terrychenism** [comment]
Here is the result generated by DecoupledNet_Full_anno model.
![qq 20150625192921](https://cloud.githubusercontent.com/assets/3353054/8369558/93512e2a-1b70-11e5-890a-a43abb2f78f4.jpg)
I have checked the final segment mask (21 channels), 1st channel is background mask which probability almost ~0.99, for 20 channel is "train" class which probability almost ~0.01. So, max(cropped_score_map, [], 3) will lead to almost "1" matrix (segmask), and get above result(almost black), is there any problems of softmax calculation?

**Comment by terrychenism** [comment]
I have normalized softmax score, by change score_map(:,:,j+1)  = softmax_score(:,:,2); to A = softmax_score(:,:,2); score_map(:,:,j+1) = A / norm(A);  which could generate reasonable result, but I think there also exist some issues.

**Comment by HyeonwooNoh** [comment]
I guess I don't understand the issue completely. Are you having trouble reproducing our reported result (http://arxiv.org/abs/1506.04924) or suggesting a better way to merge each class-specific segmentation results?

**Comment by terrychenism** [comment]
Actually I have trouble to reproduce your result. I have tested several image, most results are not good. Could you double check Softmax score in your inference code?

**Comment by HyeonwooNoh** [comment]
I checked whether this code could reproduce reported results based on VOC2012 Test score. I tested generated segmentation results on Pascal VOC 2012 set which can be obtained by running "run_demo.m". I submitted generated results on Pascal VOC evaluation server(http://host.robots.ox.ac.uk:8080/) and obtained expected score. Have you submitted the generated result to evaluation server?

**Comment by terrychenism** [comment]
I haven't submit result on PASCAL server. I have sent you some results by email. You can check it.
